Failure referencing a previous staged image in FROM clause

### Description
I'm trying to use this plugin to run a multi-stage build using a Dockerfile

Initiating the build I receive this error:
```
[ERROR] DOCKER> Unable to pull 'graalvm:latest' : {"message":"pull access denied for graalvm, repository does not exist or may require 'docker login': denied: requested access to the resource is denied"} (Not Found: 404) [{"message":"pull access denied for graalvm, repository does not exist or may require 'docker login': denied: requested access to the resource is denied"} (Not Found: 404)]

```

### Info

* d-m-p version : 0.31.0
* Maven version (`mvn -v`) : 3.6.1
```

```
* Docker version : 19.03.2
* If it's a bug, how to reproduce :
```
In my Dockerfile I have: 
FROM oracle/graalvm-ce:19.2.0 AS graalvm
...
FROM graalvm
```
In my POM I have:
```
                        <plugin>
                            <groupId>io.fabric8</groupId>
                            <artifactId>docker-maven-plugin</artifactId>
                            <version>${docker-maven-plugin.version}</version>
                            <configuration>
                                <verbose>true</verbose>
                                <images>
                                    <image>
                                        <alias>service</alias>
                                        <name>${image.namespace}/${image.name}:${image.tag}</name>
                                        <build>
                                            <noCache>false</noCache>
                                            <optimise>true</optimise>
                                            <tags>
                                                <tag>${project.version}</tag>
                                            </tags>
                                            <contextDir>${project.build.directory}/context</contextDir>
                                            <dockerFile>${project.basedir}/src/main/docker/Dockerfile</dockerFile>
                                        </build>
                                    </image>
                                </images>
                            </configuration>
                        </plugin>

```

--- a/doc/changelog.md
+++ b/doc/changelog.md
@@ -7,6 +7,7 @@
   - Treat bridged and default network mode the same #1234
   - Fix NPE when cacheFrom is missing from config #1274
   - Fix healthy option regression introduced in 0.25.0 #1279
+  - Failure referencing a previous staged image in FROM clause #1264
 
 * **0.31.0** (2019-08-10)
   - Fix test cases on Windows ([#1220](https://github.com/fabric8io/docker-maven-plugin/issues/1220))

--- a/src/main/java/io/fabric8/maven/docker/util/DockerFileUtil.java
+++ b/src/main/java/io/fabric8/maven/docker/util/DockerFileUtil.java
@@ -17,11 +17,14 @@ package io.fabric8.maven.docker.util;/*
 
 import java.io.*;
 import java.util.ArrayList;
-import java.util.LinkedList;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
+import java.util.stream.Collectors;
 
 import com.google.gson.Gson;
 import com.google.gson.JsonObject;
@@ -52,13 +55,21 @@ public class DockerFileUtil {
      */
     public static List<String> extractBaseImages(File dockerFile, FixedStringSearchInterpolator interpolator) throws IOException {
         List<String[]> fromLines = extractLines(dockerFile, "FROM", interpolator);
-        LinkedList<String> result = new LinkedList<>();
+        Set<String> result = new LinkedHashSet<>();
+        Set<String> fromAlias = new HashSet<>();
         for (String[] fromLine :  fromLines) {
             if (fromLine.length > 1) {
-                result.add(fromLine[1]);
+                if (fromLine.length == 2) { // FROM image:tag use case
+                    result.add(fromLine[1]);
+                } else if (fromLine.length == 4) { // FROM image:tag AS alias use case
+                    if (!fromAlias.contains(fromLine[1])) {
+                        result.add(fromLine[1]);
+                    }
+                    fromAlias.add(fromLine[3]);
+                }
             }
         }
-        return result;
+        return result.stream().collect(Collectors.toList());
     }
 
     /**

--- a/src/test/java/io/fabric8/maven/docker/util/DockerFileUtilTest.java
+++ b/src/test/java/io/fabric8/maven/docker/util/DockerFileUtilTest.java
@@ -59,6 +59,27 @@ public class DockerFileUtilTest {
         assertEquals(false, fromClauses.hasNext());
     }
 
+    @Test
+    public void testMultiStageNamed() throws Exception {
+        File toTest = copyToTempDir("Dockerfile_multi_stage_named_build_stages");
+        Iterator<String> fromClauses = DockerFileUtil.extractBaseImages(
+                toTest, FixedStringSearchInterpolator.create()).iterator();
+
+        assertEquals("fabric8/s2i-java", fromClauses.next());
+        assertEquals(false, fromClauses.hasNext());
+    }
+
+    @Test
+    public void testMultiStageNamedWithDuplicates() throws Exception {
+        File toTest = copyToTempDir("Dockerfile_multi_stage_named_redundant_build_stages");
+        Iterator<String> fromClauses = DockerFileUtil.extractBaseImages(
+                toTest, FixedStringSearchInterpolator.create()).iterator();
+
+        assertEquals("centos", fromClauses.next());
+        assertEquals(false, fromClauses.hasNext());
+
+    }
+
     private File copyToTempDir(String resource) throws IOException {
         File dir = Files.createTempDirectory("d-m-p").toFile();
         File ret = new File(dir, "Dockerfile");

--- a/src/test/resources/io/fabric8/maven/docker/util/Dockerfile_multi_stage_named_build_stages
+++ b/src/test/resources/io/fabric8/maven/docker/util/Dockerfile_multi_stage_named_build_stages
@@ -1,0 +1,12 @@
+# Dockerfile with multi-stage builds with named stages
+FROM fabric8/s2i-java as BUILD
+
+RUN ls -la /
+
+FROM BUILD as DEVELOPMENT
+
+RUN ls -la /
+
+FROM BUILD AS DEPOYLABLE
+
+RUN ls -la /

--- a/src/test/resources/io/fabric8/maven/docker/util/Dockerfile_multi_stage_named_redundant_build_stages
+++ b/src/test/resources/io/fabric8/maven/docker/util/Dockerfile_multi_stage_named_redundant_build_stages
@@ -1,0 +1,9 @@
+FROM centos AS first
+
+RUN ls -la /
+
+FROM centos as second
+
+RUN ls -la /
+
+FROM centos

